torch/lib/c10d/TCPStore.cpp

```diff
@@ -700,92 +888,74 @@ void TCPStore::waitForWorkers() {
-  tcputil::sendValue<QueryType>(storeSocket_, QueryType::SET);
-  tcputil::sendString(storeSocket_, regKey, true);
-  tcputil::sendVector<uint8_t>(storeSocket_, data);
+  client_->sendCommandForKey(QueryType::SET, keyPrefix_ + key);
+  client_->sendBits(data);
 }
 
 std::vector<uint8_t> TCPStore::compareSet(
     const std::string& key,
     const std::vector<uint8_t>& expectedValue,
     const std::vector<uint8_t>& desiredValue) {
-  std::string regKey = regularPrefix_ + key;
-  tcputil::sendValue<QueryType>(storeSocket_, QueryType::COMPARE_SET);
-  tcputil::sendString(storeSocket_, regKey, true);
-  tcputil::sendVector<uint8_t>(storeSocket_, expectedValue);
-  tcputil::sendVector<uint8_t>(storeSocket_, desiredValue);
-  return tcputil::recvVector<uint8_t>(storeSocket_);
+  client_->sendCommandForKey(QueryType::COMPARE_SET, keyPrefix_ + key);
+  client_->sendBits(expectedValue);
+  client_->sendBits(desiredValue);
+
+  return client_->receiveBits();
 }
 
 std::vector<uint8_t> TCPStore::get(const std::string& key) {
-  std::string regKey = regularPrefix_ + key;
-  return getHelper_(regKey);
+  return doGet(keyPrefix_ + key);
 }
 
-std::vector<uint8_t> TCPStore::getHelper_(const std::string& key) {
-  waitHelper_({key}, timeout_);
-  tcputil::sendValue<QueryType>(storeSocket_, QueryType::GET);
-  tcputil::sendString(storeSocket_, key);
-  return tcputil::recvVector<uint8_t>(storeSocket_);
+std::vector<uint8_t> TCPStore::doGet(const std::string& key) {
+  doWait(key, timeout_);
+  client_->sendCommandForKey(QueryType::GET, key);
+  return client_->receiveBits();
 }
 
 int64_t TCPStore::add(const std::string& key, int64_t value) {
-  std::string regKey = regularPrefix_ + key;
-  return addHelper_(regKey, value);
+  return incrementValueBy(keyPrefix_ + key, value);
 }
 
 bool TCPStore::deleteKey(const std::string& key) {
-  std::string regKey = regularPrefix_ + key;
-  tcputil::sendValue<QueryType>(storeSocket_, QueryType::DELETE_KEY);
-  tcputil::sendString(storeSocket_, regKey);
-  auto numDeleted = tcputil::recvValue<int64_t>(storeSocket_);
-  return (numDeleted == 1);
+  client_->sendCommandForKey(QueryType::DELETE_KEY, keyPrefix_ + key);
+  auto numDeleted = client_->receiveValue<std::int64_t>();
+  return numDeleted == 1;
 }
 
 void TCPStore::watchKey(const std::string& key, WatchKeyCallback callback) {
-  // Only allow one thread to perform watchKey() at a time
-  const std::lock_guard<std::mutex> watchKeyLock(watchKeyMutex_);
-
-  // Register callback with TCPStoreMasterDaemon to call TCPStoreWorkerDaemon on
-  // key change
-  std::string regKey = regularPrefix_ + key;
-  tcpStoreWorkerDaemon_->setCallback(regKey, callback);
-  tcputil::sendValue<QueryType>(listenSocket_, QueryType::WATCH_KEY);
-  tcputil::sendString(listenSocket_, regKey);
-
-  // Block until callback has been registered successfully
-  tcpStoreWorkerDaemon_->waitForCallbackRegistration();
+  callback_client_->setCallback(keyPrefix_ + key, callback);
```

Comment:
@H-Huang do the changes on watchKey and `TCPStoreMasterDaemon` / `TCPStoreWorkerDaemon` classes look OK to you?

Reply:
watchKey refactoring looks okay, will leave feedback for TCPStoreMasterDaemon / TCPStoreWorkerDaemon in separate comment